File: 0852-peak-index-in-a-mountain-array/0852-peak-index-in-a-mountain-array.java
class Solution {
    public int peakIndexInMountainArray(int[] a) {

        int l = 0, r = a.length - 1;

        while (l <= r) {
            int m = (l + r) >>> 1;

            if (a[m] > a[m-1] && a[m+1] < a[m]) return m;

            if (a[m] < a[m+1]) l = m;
            else r = m;

        }
        return -1;
    }
}

File: 0852-peak-index-in-a-mountain-array/0852-peak-index-in-a-mountain-array.py
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        n = len(arr)
        l, r = 1, n-2
        while l <= r:
            m = (l + r) >> 1

            if (arr[m-1] < arr[m] and arr[m] > arr[m+1]):
                return m
            elif arr[m + 1] > arr[m]:
                l = m + 1
            else:
                r = m - 1 



        return -1 